minio/minio
# Update Azure Gateway to azure-storage-blob SDK

## Description

This change updates the MinIO Azure Gateway to use the new Azure Storage SDK: [azure-storage-blob-go/azblob](https://github.com/Azure/azure-storage-blob-go).

## Motivation and Context

The currently used Azure Storage SDK [azure-sdk-for-go/storage](https://github.com/Azure/azure-sdk-for-go) has been in [maintenance-only mode](https://github.com/Azure/azure-sdk-for-go/blob/0820c05/storage/README.md#azure-storage-sdk-for-go-preview) since February 2018 and will be deprecated in the future.

Additionally, there are several limitations with the SDK, including lack of support for containerized storage deployments like [Azurite](https://github.com/Azure/Azurite) or [Azure IoT Edge storage](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-store-data-blob). These limitations currently make it hard to work with the MinIO Azure Gateway in disconnected scenarios, e.g. for local development or local integration testing.

Besides migrating away from an SDK that is no longer actively developed and adding support for containerized storage deployment backends, this change also enables us to in many cases combine previously separate PutBlob, SetMetadata and SetProperties requests into a single call which should slightly improve performance.

## How to test this PR?

First, build the code from this change into a Docker image, e.g. via `TAG=minio/minio make docker`. Then, execute the mint tests against the updated Azure gateway using Azurite as the backend storage implementation and observe that [all the tests pass](https://user-images.githubusercontent.com/1086421/69164268-ae719f00-0abd-11ea-9b8f-05dc8442c7f1.png):

```bash
MINIO_ACCESS_KEY="devstoreaccount1"
MINIO_SECRET_KEY="Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw=="

docker network create minio

docker run -d --rm --network minio --name azurite mcr.microsoft.com/azure-storage/azurite azurite-blob --blobHost 0.0.0.0

docker run -d --rm --network minio --name minio -e MINIO_ACCESS_KEY=$MINIO_ACCESS_KEY -e MINIO_SECRET_KEY=$MINIO_SECRET_KEY minio/minio gateway azure http://azurite:10000

docker run --network minio -e SERVER_ENDPOINT=minio:9000 -e ACCESS_KEY=$MINIO_ACCESS_KEY -e SECRET_KEY=$MINIO_SECRET_KEY -e ENABLE_HTTPS=0 minio/mint
```

## Types of changes


## Checklist:

- [ ] Fixes a regression
- [ ] Functional tests needed
